File: 2024/day1/solution.py
from typing import List

from utils import pyutils as adventutils


def partOne(arr1: List[int], arr2: List[int]) -> int:
    totalDistance = 0

    for i in range(len(arr1)):
        totalDistance += abs(arr1[i] - arr2[i])

    return totalDistance


def partTwo(arr1: List[int], arr2: List[int]) -> int:
    similarityScore = 0

    for i in range(len(arr1)):
        countInArr2 = arr2.count(arr1[i])
        if countInArr2 > 0:
            similarityScore += arr1[i] * countInArr2

    return similarityScore


def getSortedListsFromInput(input: str) -> List[List[int]]:
    arr1, arr2 = [], []

    for line in adventutils.file_contest_as_list_of_lines(input):
        ids = line.split("   ")
        arr1.append(int(ids[0]))
        arr2.append(int(ids[1]))

    arr1.sort()
    arr2.sort()

    return arr1, arr2


arr1, arr2 = getSortedListsFromInput("./day1/input")
print("total distance (part1): ", partOne(arr1, arr2))
print("similarity score (part2): ", partTwo(arr1, arr2))

File: 2024/day1/solution.ts
import { getFileContent } from "../utils/tsutils";

const content = await getFileContent("input");

const array1: number[] = [];
const array2: number[] = [];

for (const pair of content.split("\r\n").map(line => line.split("   "))) {
  array1.push(Number(pair[0]));
  array2.push(Number(pair[1])); 
}

array1.sort((a, b) => a - b);
array2.sort((a, b) => a - b);

console.log("total distance (part 1): ", partOne(array1, array2));
console.log("similarity score (part 2): ", partTwo(array1, array2));

function partOne(array1: number[], array2: number[]): number {
  let totalDistance = 0;
  for (let i = 0; i < array1.length; i++) {
    totalDistance += Math.abs(array1[i] - array2[i]);
  }

  return totalDistance;
}

function partTwo(array1: number[], array2: number[]): number {
  let similarityScore = 0;
  for (let i = 0; i < array1.length; i++) {
    const countInArrayTwo = array2.filter(num => num === array1[i]).length;
    if (countInArrayTwo > 0) {
      similarityScore += countInArrayTwo * array1[i];
    }
  }

  return similarityScore;
}